controllers/volumereplicationgroup_controller.go

```diff
@@ -1357,49 +1357,82 @@ func (v *VRGInstance) validateVRStatus(volRep *volrep.VolumeReplication, state r
+	conditionMet, msg := isVRConditionMet(volRep, volrepController.ConditionCompleted, metav1.ConditionTrue)
+	if !conditionMet {
+		v.updateProtectedPVCConditionHelper(volRep.Name, PVCError, msg,
+			fmt.Sprintf("VolumeReplication resource for pvc not %s to %s", action, stateString))
 
-		msg := "Failed to get the Completed condition from VolumeReplication resource for pvc."
-		v.updateProtectedPVCCondition(volRep.Name, PVCError, msg)
+		return
+	}
 
-		return nil
+	// if primary, all checks are completed
+	if state == ramendrv1alpha1.Primary {
+		msg = "VolumeReplication resource for the pvc is replicating"
+		v.updateProtectedPVCCondition(volRep.Name, PVCReplicating, msg)
+
+		return
 	}
 
-	if volRepCondition.Status == metav1.ConditionFalse {
-		v.log.Info(fmt.Sprintf("VolumeReplication resource is not %s (%s)", action, volRep.Name))
+	// it should be resyncing, if secondary
+	conditionMet, msg = isVRConditionMet(volRep, volrepController.ConditionResyncing, metav1.ConditionTrue)
+	if !conditionMet {
+		v.updateProtectedPVCConditionHelper(volRep.Name, PVCError, msg,
+			"VolumeReplication resource for pvc not resyncing as Secondary")
 
-		msg := fmt.Sprintf("VolumeReplication resource for pvc not %s to %s", action, stateString)
-		v.updateProtectedPVCCondition(volRep.Name, PVCError, msg)
+		return
+	}
 
-		return nil
+	// it should not be degraded, if secondary
+	conditionMet, msg = isVRConditionMet(volRep, volrepController.ConditionDegraded, metav1.ConditionFalse)
+	if !conditionMet {
+		v.updateProtectedPVCConditionHelper(volRep.Name, PVCError, msg,
+			"VolumeReplication resource for pvc is degraded")
+
+		return
+	}
+
+	msg = "VolumeReplication resource for the pvc is replicating"
+	v.updateProtectedPVCCondition(volRep.Name, PVCReplicating, msg)
+}
+
+func isVRConditionMet(volRep *volrep.VolumeReplication,
+	conditionType string,
+	desiredStatus metav1.ConditionStatus) (bool, string) {
+	volRepCondition := findCondition(volRep.Status.Conditions, conditionType)
+	if volRepCondition == nil {
+		msg := fmt.Sprintf("Failed to get the %s condition from status of VolumeReplication resource.", conditionType)
+
+		return false, msg
+	}
+
+	if volRep.Generation != volRepCondition.ObservedGeneration {
+		msg := fmt.Sprintf("Stale generation for condition %s from status of VolumeReplication resource.", conditionType)
+
+		return false, msg
 	}
 
 	if volRepCondition.Status == metav1.ConditionUnknown {
-		v.log.Info(fmt.Sprintf("VolumeReplication resource is not %s (%s). ConditionStatus unknown",
-			action, volRep.Name))
+		msg := fmt.Sprintf("Unknown status for condition %s from status of VolumeReplication resource.", conditionType)
 
-		msg := fmt.Sprintf("VolumeReplication resource for pvc not %s to %s. ConditionStatus unknown",
-			action, stateString)
-		v.updateProtectedPVCCondition(volRep.Name, PVCErrorUnknown, msg)
+		return false, msg
+	}
 
-		return nil
+	if volRepCondition.Status != desiredStatus {
+		return false, ""
```

Comment:
Should we return some message here indicating what the error is?

Reply:
The message changes based on type and condition that failed or is not met yet, hence letting the function specify a default message when no message is returned from this function, to distinguish the various cases.

Reply:
ACK